source/common/network/listen_socket_impl.cc

```diff
@@ -37,6 +38,22 @@ TcpListenSocket::TcpListenSocket(const Address::InstanceConstSharedPtr& address,
   int rc = setsockopt(fd_, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
   RELEASE_ASSERT(rc != -1);
 
+#if defined SOL_IP && defined IP_TRANSPARENT
+  // Set the SOL_IP/IP_TRANSPARENT socket option to support running with iptables TPROXY target
+  // instead of REDIRECT to get both the original source and destination addresses of the
+  // redirected socket.
+  on = 1;
+  rc = setsockopt(fd_, SOL_IP, IP_TRANSPARENT, &on, sizeof(on));
```

Comment:
AFAIK, this should be `IP_TRANSPARENT` for IPv4 and `IPV6_TRANSPARENT` for IPv6.

Reply:
Fixed in upcoming commit.

Reply:
Note that this won't work for IPv6 when building using Envoy's Docker image (Ubuntu 16.04.3 with libc6-dev 2.23-0ubuntu10) because `IPV6_TRANSPARENT` is not defined in that image:
```
$ ./ci/run_envoy_docker.sh 'grep IPV6_TRANSPARENT /usr/include/x86_64-linux-gnu/bits/in.h'
```
returns nothing. 

On Ubuntu 17.10 with libc6-dev 2.26-0ubuntu2.1, this macro is defined as expected:
```
$ grep IPV6_TRANSPARENT /usr/include/x86_64-linux-gnu/bits/in.h
#define IPV6_TRANSPARENT	75
```